http/cookies.lua

```diff
@@ -0,0 +1,324 @@
+local http_patts = require "lpeg_patterns.http"
+local util = require "http.util"
+local psl = require "psl"
+
+local function parse_set_cookie(text_cookie, request, time)
+	local domain = request.headers:get("host")
+	local key, value, matched_cookie = assert(http_patts.Set_Cookie:match(
+		text_cookie))
+	-- allow custom time for specs to not error
+	time = time or os.time()
+	local cookie = {
+		creation = time;
+		last_access = time;
+		persistent =
+			not not (matched_cookie.expires or matched_cookie["max-age"]);
+		domain = matched_cookie.domain or domain;
+		path = matched_cookie.path or request.headers:get(":path");
+		secure = matched_cookie.secure or false;
+		http_only = matched_cookie.httponly or false;
+		key = key;
+		value = value;
+		host_only = not not matched_cookie.domain
+	}
+	cookie.indexname = cookie.domain .."|".. cookie.path .."|".. cookie.key
+	if matched_cookie["max-age"] then
+		cookie.expires = time + tonumber(matched_cookie["max-age"])
+	else -- luacheck: ignore
+		-- ::TODO:: make use of `expires` cookie value
+	end
+	local same_site do
+		if matched_cookie.samesite == true then
+			same_site = "lax"
+		else
+			same_site = matched_cookie.samesite
+		end
+	end
+	cookie.same_site = same_site
+	return cookie
+end
+
+local function bake_cookie(data)
+	assert(data.key, "cookie must contain a `key` field")
+	assert(data.value, "cookie must contain a `value` field")
+	local cookie = {tostring(data.key) .. "=" .. tostring(data.value)}
+	if data.expires then
+		cookie[#cookie + 1] = "Expires=" .. util.imf_date(data.expires)
+	end
+	if data.max_age then
+		cookie[#cookie + 1] = "Max-Age=" .. tostring(math.floor(data.max_age))
+	end
+	if data.domain then
+		cookie[#cookie + 1] = "Domain=" .. data.domain
+	end
+	if data.path then
+		cookie[#cookie + 1] = "Path=" .. data.path
+	end
+	if data.secure then
+		cookie[#cookie + 1] = "Secure"
+	end
+	if data.http_only then
+		cookie[#cookie + 1] = "HttpOnly"
+	end
+	if data.same_site then
+		if data.same_site == true then
+			data.same_site = "Strict"
+		end
+		cookie[#cookie + 1] = "SameSite=" .. data.same_site:gsub("^.",
+			string.upper)
```

Comment:
Why title case here? What happens if it's lower case?
SameSite seems to only be allowed to be "Strict" or "Lax", so we can hard code the whole string (and assert that `data.same_site` is a boolean)